arroyo/processing/strategies/produce.py

```diff
@@ -0,0 +1,86 @@
+import logging
+import time
+from collections import deque
+from concurrent.futures import Future
+from typing import Deque, Optional, Tuple
+
+from arroyo.backends.abstract import Producer
+from arroyo.processing.strategies.abstract import MessageRejected, ProcessingStrategy
+from arroyo.types import Commit, Message, Position, Topic, TPayload
+
+logger = logging.getLogger(__name__)
+
+
+class ProduceAndCommit(ProcessingStrategy[TPayload]):
+    """
+    For each message received, produces a message to the topic being passed in.
+    Waits on the futures and only once completed, commits the offset of the message.
+    """
+
+    def __init__(
+        self,
+        producer: Producer[TPayload],
+        topic: Topic,
+        commit: Commit,
+        max_buffer_size: int = 10000,
+    ):
+        self.__producer = producer
+        self.__topic = topic
+        self.__commit = commit
+        self.__max_buffer_size = max_buffer_size
+
+        self.__queue: Deque[
+            Tuple[Message[TPayload], Future[Message[TPayload]]]
+        ] = deque()
+
+        self.__closed = False
+
+    def poll(self) -> None:
+        while self.__queue:
+            message, future = self.__queue[0]
+
+            if not future.done():
+                break
+
+            self.__queue.popleft()
+
+            self.__commit(
+                {message.partition: Position(message.next_offset, message.timestamp)}
+            )
+
+    def submit(self, message: Message[TPayload]) -> None:
+        assert not self.__closed
+
+        if len(self.__queue) >= self.__max_buffer_size:
+            raise MessageRejected
```

Comment:
Knowing that, as of now, `MessageRejected` are often not properly managed by a step that send messages to this one and that when the Processor receives these back pressure events, it goes into a tight cpu loop waiting, wouldn't you prefer preventing this occurrence from happening at all, and waiting on the queue in the `poll` method instead  so that this condition is never verified? This until we make the `MessageRejected` process better.

I'd be ok either way though.

Reply:
Though if you keep `MessageRejected` around, in this step you will have to fix the parallel transform step as it does not handle those exceptions:
https://github.com/getsentry/arroyo/blob/main/arroyo/processing/strategies/streaming/transform.py#L397-L403.
Without fixing that, this step would not be usable after a parallel transform.

Reply:
Yes, there an existing TODO in the parallel transform but that is not particularly relevant to this strategy. `MessageRejected` is the only proper Arroyo way of doing back pressure, I don't think we should do a workaround because of the bug in the other strategy.

Reply:
I think it is relevant to this strategy as whoever creates a consumer cannot use a parallel transform strategy before this one. 
Could you call that out in the docstring? IT will be easy to miss that case when building a consumer.

Reply:
Ok. I added it in both the docstring of this class as well as parallel tranform